### PR TITLE
introduce  gradient update handler to the  base estimator

### PR DESCRIPTION
**Pull request by liuzh91**
## Description ##
This change add default gradient update handler to the base estimator class. The purpose of this change is to decouple the forward/backward computation with the gradient apply process. In some use cases such as gradient accumulation, gradient updates are not executed during each training batch.  See issue description #16869.

### CODE REVIEW

**Comment by ptrendx** [comment]
@leezu What is the rationale behind marking this PR as 1.6? This looks like adding a new functionality.

File: tests/python/unittest/test_gluon_estimator.py [hunk 1]
```diff
@@ -332,39 +332,41 @@ def test_default_handlers():
...
     est = Estimator(net=net,
                     loss=loss,
                     metrics=train_acc,
                     trainer=trainer,
                     context=ctx)
-    # no handler(all default handlers), no warning
+    # no handler except gradient update handler (all default handlers), no warning
     with warnings.catch_warnings(record=True) as w:
-        est.fit(train_data=train_data, epochs=num_epochs)
+        est.fit(train_data=train_data, epochs=num_epochs, event_handlers=[gradient_update])
```
Comment (roywei) [inline tests/python/unittest/test_gluon_estimator.py]: you don't need to add `gradient_update` here right? as it's added as a default handler?

Comment (liuzh91) [reply tests/python/unittest/test_gluon_estimator.py]: My mistake. I'll remove this handler here.

**Comment by leezu** [comment]
@ptrendx the estimator API is currently experimental and is not yet up to more complex real-world use-cases. @liuzh91 is currently converting GluonNLP training scripts to make use of the Estimator API and ran into several remaining shortcomings of the API.
We shouldn't release 1.6 with a known broken version of Estimator *if fixes are available*. Thus if this PR is approved and merged to master in time for the 1.6 release, it may be good to backport. Given the scope of the estimator API and the current shortcomings we can consider these PRs as bugfixes. If we don't fix the API with the 1.6 release, we may be bound to the broken API due to backwards compatibility commitments.

Do you think backporting the fix would be reasonable or do you have concerns? When do you plan to tag a release candidate?

These extends the fixes commited started November 2019 https://github.com/apache/incubator-mxnet/commits/master/python/mxnet/gluon/contrib/estimator.

File: python/mxnet/gluon/contrib/estimator/estimator.py [hunk 2]
```diff
@@ -295,13 +295,11 @@ def fit_batch(self, train_batch, batch_axis=0):
 
         with autograd.record():
             pred = [self.net(x) for x in data]
-            loss = [self.loss(y_hat, y) for y_hat, y in zip(pred, label)]
+            loss = [self.loss(y_hat, y) / batch_size for y_hat, y in zip(pred, label)]
```
[the remark this answers is not in the document]
Comment (liuzh91) [reply python/mxnet/gluon/contrib/estimator/estimator.py]: I'll look into this problem.

**Comment by liuzh91** [comment]
> Thank you for the improvement! 2 concerns.
> Also could you point an example which require custom gradient handler? (gradient clipping or aggregation)

Thank u for the review. 

For the gradient update example, one use case of using gradient accumulation appears when training a transformer. (https://github.com/dmlc/gluon-nlp/blob/master/scripts/machine_translation/train_transformer.py#L320) Because the size of parameters in the transformer network is too large, we can compute gradient for a small batch of data examples. In this case, the gradient is updated periodically on the weight parameters.

File: python/mxnet/gluon/contrib/estimator/event_handler.py [hunk 3]
```diff
@@ -130,13 +130,15 @@ class MetricHandler(EpochBegin, BatchEnd):
     ----------
     train_metrics : List of EvalMetrics
         Training metrics to be updated at batch end.
+    priority : scalar
+        Priority level of the MetricHandler
```
Comment (leezu) [inline python/mxnet/gluon/contrib/estimator/event_handler.py]: Can you explain that the priority is used for ascending sorting of the handlers?

File: python/mxnet/gluon/contrib/estimator/event_handler.py [hunk 4]
```diff
@@ -176,14 +178,17 @@ class ValidationHandler(TrainBegin, BatchEnd, EpochEnd):
     batch_period : int, default None
         How often to run validation at batch end, by default
         :py:class:`ValidationHandler` does not validate at batch end.
+    priority: scalar, default -1000
+        Priority level of the ValidataionHandler
```
Comment (leezu) [inline python/mxnet/gluon/contrib/estimator/event_handler.py]: Typo

File: python/mxnet/gluon/contrib/estimator/event_handler.py [hunk 5]
```diff
@@ -235,14 +240,17 @@ class LoggingHandler(TrainBegin, TrainEnd, EpochBegin, EpochEnd, BatchBegin, Bat
         Training metrics to be logged, logged at batch end, epoch end, train end.
     val_metrics : list of EvalMetrics
         Validation metrics to be logged, logged at epoch end, train end.
+    priority : scalar, default np.Inf
+        Priority level of the LoggingHandler
```
Comment (leezu) [inline python/mxnet/gluon/contrib/estimator/event_handler.py]: Explanation of the priority level would be helpful for users.

File: python/mxnet/gluon/contrib/estimator/event_handler.py [hunk 6]
```diff
@@ -706,3 +714,29 @@ def train_end(self, estimator, *args, **kwargs):
...
+
+class GradientUpdateHandler(BatchEnd):
+    """Gradient Update Handler that apply gradients on network weights
+
+    :py:class:`GradientUpdateHandler` takes the priority level. It updates weight parameters
+    at the end of each batch
+
+    Parameters
+    ----------
+    priority : scalar, default -np.Inf
```
Comment (leezu) [inline python/mxnet/gluon/contrib/estimator/event_handler.py]: Should a integer be used instead of `-np.Inf`? Maybe there are some use-cases where a handler should be run prior to the gradient handler?

File: python/mxnet/gluon/contrib/estimator/event_handler.py (hunk 3 above)
Comment (liuzh91) [reply python/mxnet/gluon/contrib/estimator/event_handler.py]: I have fixed these documentation issues.

**Comment by liuzh91** [comment]
>     1. Does this require the user to specify grad_req = 'add' for gradient accumulation? When is the gradient reset to zeros?
> 
>     2. Now GradientUpdateHandler is added to the list of default handlers. If users explicitly create estimator with `event_handlers=[metric, logging]`, will `GradientUpdateHandler` still be included in the estimator and performing updates correctly?

1. The gradient update handler is for general gradient update. When we implement gradient accumulation handler in the future, we will inherit this class.

2. Yes, it is added as a default handler. Please refer to (https://github.com/apache/incubator-mxnet/blob/5878451c4b476d8771d93000ff5d51b81071997d/tests/python/unittest/test_gluon_estimator.py#L369).
